<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Описание данных</a></span></li></ul></li><li><span><a href="#Шаг-1" data-toc-modified-id="Шаг-1-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-1.0.1"><span class="toc-item-num">1.0.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li></ul></li><li><span><a href="#Задача-1" data-toc-modified-id="Задача-1-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Задача 1</a></span><ul class="toc-item"><li><span><a href="#Ответ" data-toc-modified-id="Ответ-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Ответ</a></span></li></ul></li><li><span><a href="#Задача-2" data-toc-modified-id="Задача-2-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задача 2</a></span><ul class="toc-item"><li><span><a href="#Ответ" data-toc-modified-id="Ответ-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Ответ</a></span></li></ul></li><li><span><a href="#Задача-3" data-toc-modified-id="Задача-3-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задача 3</a></span><ul class="toc-item"><li><span><a href="#Ответ" data-toc-modified-id="Ответ-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Ответ</a></span></li></ul></li><li><span><a href="#Задача-4" data-toc-modified-id="Задача-4-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задача 4</a></span><ul class="toc-item"><li><span><a href="#Ответ" data-toc-modified-id="Ответ-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Ответ</a></span></li></ul></li><li><span><a href="#Задача-5" data-toc-modified-id="Задача-5-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Задача 5</a></span><ul class="toc-item"><li><span><a href="#Ответ" data-toc-modified-id="Ответ-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Ответ</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></div>

# SQL
Крупный сервис для чтения книг по подписке. 

Первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Описание даннх
### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Шаг 1 
Прочитаь таблицы

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
query = (
''' select *
    from books
''')
books = pd.io.sql.read_sql(query, con = engine)
books.head(5)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [4]:
query = (
''' select *
    from authors
''')
authors = pd.io.sql.read_sql(query, con = engine)
authors.head(5)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [6]:
query = (
''' select *
    from publishers
    limit 5
''')
pd.io.sql.read_sql(query, con = engine)


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [7]:
query = (
''' select *
    from ratings

''')
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.sample(5)

rating_id  book_id        username  rating
3923       3924      635         colin90       4
4083       4084      665  shermannatalie       4
5597       5598      863   kevincampbell       4
3433       3434      545    johngonzalez       5
3043       3044      481     danielchung       4

In [8]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [9]:
query = (
''' select *
    from reviews
    
''')
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.head(5)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [10]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


#### Вывод:
- Обзоров меньше, чем рейтингов
- Книг всего 1000
- авторов 636
- издательств 5

## Задача 1
Посчитайте, сколько книг вышло после 1 января 2000 года

In [11]:
query = (
'''
SELECT 
COUNT(*)
FROM books
WHERE CAST(publication_date AS date) >'2000-01-01' 
''')
pd.io.sql.read_sql(query, con = engine)


count
0    819

### Ответ
819 книг вышло после 1.01.2000

## Задача 2
Для каждой книги посчитайте количество обзоров и среднюю оценку

In [12]:
query = (
'''
select b.book_id, 
count(r.review_id) as review_count,
avg(rat.rating)
from books b
left outer join  reviews as r on b.book_id = r.book_id
left outer join  ratings as rat on r.book_id = rat.book_id
group by b.book_id
''')
pd.io.sql.read_sql(query, con = engine)

book_id  review_count       avg
0        652             4  4.500000
1        273             4  4.500000
2         51            60  4.250000
3        951             4  4.000000
4        839            28  4.285714
..       ...           ...       ...
995       64            52  4.230769
996       55             4  5.000000
997      148            21  3.428571
998      790             4  3.500000
999      828             4  3.000000

[1000 rows x 3 columns]

### Ответ
отображена таблица, в которой есть средняя оценка по каждой книге и кол-во обзоров

## Задача 3
Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [13]:
query = (
'''
select p.publisher, count(*) count_books 
from books as b
INNER JOIN publishers as p on b.publisher_id = p.publisher_id
where b.num_pages > 50
group by p.publisher
order by count(*) desc
limit 1
''')
pd.io.sql.read_sql(query, con = engine)

publisher  count_books
0  Penguin Books           42

### Ответ
издательство Penguin Books

## Задача 4
Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [14]:
query = (
'''
select a.author, avg(rat.rating)
from books as b
INNER JOIN authors as a on b.author_id = a.author_id
INNER JOIN ratings as rat on b.book_id = rat.book_id
group by author
having count(*) >= 50
limit 1
''')
pd.io.sql.read_sql(query, con = engine)

author       avg
0  J.K. Rowling/Mary GrandPré  4.288462

### Ответ
автоп J.K. Rowling/Mary GrandPré имеет самую высокую среднюю оценку среди книг с 50 и более оценками

## Задача 5
Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [15]:
#спомогательные таблицы для себя
query = '''
            SELECT username, COUNT(rating)
            FROM ratings
            GROUP BY username
            HAVING COUNT(rating) > 50
                 '''
pd.io.sql.read_sql(query, con = engine)

username  count
0     sfitzgerald     55
1  jennifermiller     53
2          xdavis     51
3          paul88     56
4      martinadam     56
5       richard89     55

In [16]:
query = '''
            SELECT username, COUNT(review_id)
            FROM reviews
            GROUP BY username
            
                 '''
pd.io.sql.read_sql(query, con = engine)

username  count
0         zjohnston     14
1           jnelson     21
2      lewisdesiree     19
3     patrickhudson     15
4             ulowe     16
..              ...    ...
155  shermannatalie     24
156         dmiller     18
157  charlesraymond     16
158     isaiahreyes     15
159       masonsara     14

[160 rows x 2 columns]

In [17]:
query = (
'''
WITH
rat AS (SELECT username, COUNT(rating)
            FROM ratings
            GROUP BY username
            HAVING COUNT(rating) > 50),

r AS (SELECT username, COUNT(review_id) as review_count
            FROM reviews
            GROUP BY username)
SELECT avg(review_count)
FROM rat LEFT OUTER JOIN r ON rat.username=r.username;


''')
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

### Ответ
среднее кол-во обзоров от пользователей, которые поставили более 50 оценок: 24,3

### Вывод
в рамках проекта представлено 5 таблиц с данными от книжного сервиса:
- Книг представлено всего 1000
- Авторов 636
- Издательств 5
- В таблице ratings  рейтингов проставлено больше, чем обзоров (написать обхор более трудозатратно, скорее всего связано с этим)

* Для большинства книг было написано несколько обзоров 
* Издательство "Penguin Books" выпустило 42 книги толще 50 страниц и является самым выпускаемым для такого объема
* Автор с самой высокой оценкой - Джоан Роулинг в соавторстве с Мэри Гранпрэ(иллюстратором). Вполне ожидаемо, так как книги о Гарри Поттере одни из самых популярных и любимых
* Среднее количество обзоров от пользователей, которые поставили более 50 оценок - 24. Можно предположить, что пользователям нравится  серсис и они с удовольствием им пользуются